hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/BlockInfoContiguous.java

```diff
@@ -77,9 +85,31 @@ boolean removeStorage(DatanodeStorageInfo storage) {
     setStorageInfo(dnIndex, getStorageInfo(lastNode));
     // set the last entry to null
     setStorageInfo(lastNode, null);
+    if (storage.getStorageType() == StorageType.PROVIDED
+        && !hasProvidedStorages()) {
```

Comment:
Shouldn't this already guarantee is false?

Reply:
I see is the other method which goes over each DN, nevermind.

Reply:
Can you go over the overhead added to the regular pipeline without Provided?

Reply:
I removed the check and there is no invocation in the regular pipeline now.